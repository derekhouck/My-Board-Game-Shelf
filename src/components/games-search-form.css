.games__search-fields {
  border: 0;
  display: flex;
  padding: 0;
}

.games__search-fields .form-input {
  margin-right: 1em;
}