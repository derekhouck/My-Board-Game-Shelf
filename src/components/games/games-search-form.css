.games__search {
  text-align: center;
}

.games__search-fields legend {
  text-transform: uppercase;
}

.games__search-buttons > * {
  margin: 1em;
}

@media only screen and (min-width: 768px) {
  .games__search-fields {
    display: grid;
    grid-template-areas: 
      "title input"
      "tags tags";
    justify-content: center;
  }

  .games__search-fields .form-input {
    margin-right: 1em;
  }

  .games__search-title {
    grid-area: title;
  }
  
  .games__search-tags {
    grid-area: tags;
  }
}

@media only screen  and (min-width : 1224px) {
  .games__search-fields {
    display: block;
  }

  .games__search-fields .form-input {
    margin: 0.5em;
  }

  .games__search-buttons > * {
    display: block;
    margin: 1em auto;
  }
}