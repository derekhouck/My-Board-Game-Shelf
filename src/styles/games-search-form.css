.games__search {
  text-align: center;
}

.games__search-fields {
  border: 0;
  padding: 0;
}

@media only screen and (min-width: 768px) {
  .games__search-fields {
    display: flex;
  }
  .games__search-fields .form-input {
    margin-right: 1em;
  }
}