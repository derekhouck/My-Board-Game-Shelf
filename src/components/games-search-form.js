import React from 'react';
import { connect } from 'react-redux';
import { reduxForm, Field, focus } from 'redux-form';
import { filterGames, resetFilters } from '../actions/games';
import '../styles/games-search-form.css';

import Button from './button';
import Input from './input';
import Select from './select';

export class GamesSearchForm extends React.Component {
  onSubmit(values) {
    const { searchTerm, players, tag } = values;
    const filters = {
      title: searchTerm,
      players,
      tagId: tag
    };
    return this.props.dispatch(filterGames(filters));
  }

  render() {
    const tags = this.props.tags.map(tag => ({ value: tag.id, text: tag.name }));

    return (
      <section className="games__search">
        <h2>Filter your games</h2>
        <form
          className="games__search-form"
          onSubmit={this.props.handleSubmit(values => this.onSubmit(values))}
        >
          <div className="games__search-fields">
            <Field
              component={Input}
              type="text"
              name="searchTerm"
              id="searchTerm"
              label="Game title"
              className="games__search-title"
            />
            <Field
              component={Input}
              type="number"
              name="players"
              id="players"
              label="Number of players"
            />
            <div className="form-input games__search-tags">
              <label htmlFor="tag">Tag</label>
              <Field
                component={Select}
                id="tag"
                name="tag"
                noSelectionLabel='Select a tag'
                options={tags}
              />
            </div>
          </div>
          <div className="games__search-buttons">
            <Button label="Search" type="submit" />
            <Button secondary label="Clear filters" type="reset" onClick={() => this.props.dispatch(resetFilters())} />
          </div>
        </form>
      </section>
    );
  }
}

function getUnique(arr, comp) {

  const unique = arr
    .map(e => e[comp])

    // store the keys of the unique objects
    .map((e, i, final) => final.indexOf(e) === i && i)

    // eliminate the dead keys & store unique objects
    .filter(e => arr[e]).map(e => arr[e]);

  return unique;
}

const getTags = (games) => {
  const tagArrays = games.games.map(game => game.tags);
  const mergedArray = [].concat.apply([], tagArrays);
  const tags = getUnique(mergedArray, 'id');

  return tags;
};

const mapStateToProps = state => ({
  tags: getTags(state.games)
});

export default reduxForm({
  form: 'games-search-form',
  onSubmitFail: (errors, dispatch) => dispatch(focus('registration', Object.keys(errors)[0]))
})(connect(mapStateToProps)(GamesSearchForm));