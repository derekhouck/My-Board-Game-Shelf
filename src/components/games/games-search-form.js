import React from 'react';
import { connect } from 'react-redux';
import { reduxForm, Field, focus } from 'redux-form';
import { filterGames, resetFilters } from '../../actions/games';
import './games-search-form.css';

import Button from '../atoms/button';
import Input from '../atoms/input';
import Select from '../select';
import { separateTags } from '../helpers';

export class GamesSearchForm extends React.Component {
  state = {
    mechanics: [],
    themes: []
  };

  componentDidUpdate(prevProps) {
    const { tags } = this.props;
    if (prevProps.tags !== tags) {
      const tagsObj = separateTags(tags);
      this.setState({
        mechanics: tagsObj.mechanics,
        themes: tagsObj.themes,
      })
    }
  }

  componentWillUnmount() {
    const { dispatch } = this.props;
    return dispatch(resetFilters());
  }
  onSubmit(values) {
    const { mechanics, players, searchTerm, themes } = values;
    const filters = {
      title: searchTerm,
      mechanics,
      players,
      tagId: mechanics && mechanics !== 'null' ? mechanics : themes,
      themes,
    };
    return this.props.dispatch(filterGames(filters));
  }

  render() {
    const { dispatch, handleSubmit, initialValues } = this.props;
    const { mechanics, themes } = this.state;

    return (
      <section className="games__search">
        <form
          className="games__search-form"
          onSubmit={handleSubmit(values => this.onSubmit(values))}
        >
          <fieldset className="games__search-fields">
            <legend>Filter your games</legend>
            <Field
              component={Input}
              name="searchTerm"
              id="searchTerm"
              label="Game title"
              className="games__search-title"
              placeholder={initialValues.title}
              type="text"
            />
            <Field
              component={Input}
              name="players"
              id="players"
              label="Number of players"
              placeholder={initialValues.players}
              type="number"
            />
            <div className="form-input games__search-mechanics">
              <label htmlFor="mechanics">Mechanics</label>
              <Field
                component={Select}
                id="mechanics"
                initialValue={initialValues.mechanics}
                name="mechanics"
                noSelectionLabel='Select a mechanic'
                options={
                  mechanics.map(mechanic => ({ value: mechanic.id, text: mechanic.name }))
                }
              />
            </div>
            <div className="form-input games__search-themes">
              <label htmlFor="themes">Themes</label>
              <Field
                component={Select}
                id="themes"
                initialValue={initialValues.themes}
                name="themes"
                noSelectionLabel='Select a theme'
                options={
                  themes.map(theme => ({ value: theme.id, text: theme.name }))
                }
              />
            </div>
          </fieldset>
          <div className="games__search-buttons">
            <Button label="Search" type="submit" />
            <Button secondary label="Clear filters" type="reset" onClick={() => dispatch(resetFilters())} />
          </div>
        </form>
      </section>
    );
  }
}

GamesSearchForm.defaultProps = {
  initialValues: {}
};

function getUnique(arr, comp) {
  const unique = arr
    .map(e => e[comp])
    // store the keys of the unique objects
    .map((e, i, final) => final.indexOf(e) === i && i)
    // eliminate the dead keys & store unique objects
    .filter(e => arr[e]).map(e => arr[e]);
  return unique;
}

const getTags = (games) => {
  const tagArrays = games.map(game => game.tags);
  const mergedArray = [].concat.apply([], tagArrays);
  const tags = getUnique(mergedArray, 'id');
  return tags;
};

const mapStateToProps = (state, ownProps) => {
  const { games } = ownProps;
  return {
    initialValues: state.games.filters,
    tags: getTags(games)
  }
};

export default connect(mapStateToProps)(
  reduxForm({
    enableReinitialize: true,
    form: 'games-search-form',
    onSubmitFail: (errors, dispatch) => dispatch(focus('games-search-form', Object.keys(errors)[0]))
  })(GamesSearchForm)
);