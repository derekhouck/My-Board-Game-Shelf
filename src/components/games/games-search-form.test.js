import React from 'react';
import { shallow } from 'enzyme';
import { GamesSearchForm } from './games-search-form';

describe('<GamesSearchForm />', function () {
  it('renders without crashing', () => {
    const callback = jest.fn();
    const wrapper = shallow(<GamesSearchForm tags={[]} handleSubmit={callback} />);
    expect(wrapper).toMatchSnapshot();
  });
});
