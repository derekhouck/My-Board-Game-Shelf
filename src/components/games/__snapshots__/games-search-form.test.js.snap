// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GamesSearchForm /> renders without crashing 1`] = `ShallowWrapper {}`;
